extensions/amp-call-tracking/0.1/amp-call-tracking.js

```diff
@@ -0,0 +1,98 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Layout} from '../../../src/layout';
+import {assertHttpsUrl} from '../../../src/url';
+import {urlReplacementsForDoc} from '../../../src/url-replacements';
+import {user} from '../../../src/log';
+import {xhrFor} from '../../../src/xhr';
+
+
+/**
+ * Bookkeeps all unique URL requests so that no URL is called twice.
+ * @type {!Object<!string, !Promise>}
+ */
+let cachedResponsePromises_ = {};
+
+
+/**
+ * Fetches vendor response.
+ * @param {!Window} win
+ * @param {!string} url
+ * @return {!Promise<Object>}
+ */
+function fetch_(win, url) {
+  if (!(url in cachedResponsePromises_)) {
+    cachedResponsePromises_[url] = xhrFor(win).fetchJson(url);
+  }
+  return cachedResponsePromises_[url];
+}
+
+
+/** Visible for testing. */
+export function clearResponseCache() {
+  cachedResponsePromises_ = {};
+}
+
+
+/**
+ * Implementation of `amp-call-tracking` component. See
+ * {@link ../amp-call-tracking.md} for the spec.
+ */
+export class AmpCallTracking extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {?Element} */
+    this.hyperlink_ = null;
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout === Layout.FIXED ||
+        layout === Layout.FIXED_HEIGHT ||
+        layout === Layout.RESPONSIVE ||
+        layout === Layout.CONTAINER ||
+        layout === Layout.FILL ||
+        layout === Layout.FLEX_ITEM;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.hyperlink_ = this.getRealChildren()[0];
```

Comment:
That is better for clarity, true. But:

- Querying for an `<a>` element specifically is not needed since the validator ensures that the child node will be an anchor tag.

- Checking for the `href` attribute to contain a `tel:` prefix is a little overzealous. If the client sets the default phone number incorrectly that's up to them.

I don't feel very strongly about this either way, but I thought these are important aspects to account for.